### PR TITLE
POLIO-1645 POLIO-1646  POLIO-1647  incident reports improvements

### PR DESCRIPTION
**Pull request by beygorghor**
Explain what problem this PR is resolving

Related JIRA tickets : POLIO-1645 POLIO-1646  POLIO-1647  

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] Are my typescript files well typed
- [x] New translations have been added or updated if new strings have been introduced in the frontend
- [ ] My migrations file are included
- [ ] Are there enough tests
- [x] Documentation has been included (for new feature)

## Doc
Some doc is present on plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/Modals/CreateEditIncident.tsx

## Changes

- added title and comment to incident report model + migrations
- added those field in form
- added Other option in types, not displayed in the frontend for now as POLIO-1650 ticket is hold due to the complexity of the feature
- review number inputs logic on the frontend to make it clear and match to this:
Screenshot 2024-08-22 at 12 23 26

## How to test

Open a stock variation for a country with existing data. 
Screenshot 2024-08-22 at 16 03 04

open the details for one entry and click on stock variation.
Select Incident reports tab .
Edit and create reports, check it matches the sheet
Add title and comment and save.

## Print screen / video

Screenshot 2024-08-22 at 15 58 01
Screenshot 2024-08-22 at 15 57 55
Screenshot 2024-08-22 at 15 57 49
Screenshot 2024-08-22 at 15 57 43

### CODE REVIEW

File: plugins/polio/api/vaccines/stock_management.py [hunk 1]
```diff
@@ -174,6 +175,7 @@ def get_list_of_usable_vials(self):
                 or report.stock_correction == IncidentReport.StockCorrectionChoices.VVM_REACHED_DISCARD_POINT
                 or report.stock_correction == IncidentReport.StockCorrectionChoices.UNREADABLE_LABEL
                 or report.stock_correction == IncidentReport.StockCorrectionChoices.BROKEN
+                or report.stock_correction == IncidentReport.StockCorrectionChoices.OTHER
```
Comment (quang-le) [inline plugins/polio/api/vaccines/stock_management.py]: This is another PR no?

File: plugins/polio/api/vaccines/stock_management.py [hunk 2]
```diff
@@ -261,6 +263,7 @@ def get_list_of_unusable_vials(self):
                 or report.stock_correction == IncidentReport.StockCorrectionChoices.VACCINE_EXPIRED
                 or report.stock_correction == IncidentReport.StockCorrectionChoices.VVM_REACHED_DISCARD_POINT
                 or report.stock_correction == IncidentReport.StockCorrectionChoices.UNREADABLE_LABEL
+                or report.stock_correction == IncidentReport.StockCorrectionChoices.OTHER
```
Comment (quang-le) [inline plugins/polio/api/vaccines/stock_management.py]: same same

File: plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/Modals/CreateEditIncident.tsx [hunk 3]
```diff
@@ -38,18 +115,82 @@ export const CreateEditIncident: FunctionComponent<Props> = ({
...
+            event.target.value === 'usable' ? 'unusable_vials' : 'usable_vials',
+            0,
+        );
+    };
+
+    const getInitialMovement = useCallback(() => {
+        if (!incident) return 0;
+        const movementType = incidentReportConfig[incident.stock_correction];
+        return movementType === 'inventory' ? 0 : incident.usable_vials;
+    }, [incident]);
```
Comment (quang-le) [inline plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/Modals/CreateEditIncident.tsx]: Couldn't you use `useMemo` instead, to pass the return value directly to formik?
Honest question, I can imagine that we would want the value passed to `initialValues` to be static, and not "stateful".

Alternately, I think you can move the function outside of the component:
```typescript
const getInitialMovement = (incident) => {
        if (!incident) return 0;
        const movementType = incidentReportConfig[incident.stock_correction];
        return movementType === 'inventory' ? 0 : incident.usable_vials;
    }
```

File: plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/Modals/CreateEditIncident.tsx [hunk 4]
```diff
@@ -59,6 +200,31 @@ export const CreateEditIncident: FunctionComponent<Props> = ({
...
+        movement: number,
+    ) => {
+        if (movementType === 'plainMovement') {
+            return formatMessage(MESSAGES.plainMovement, { movement });
+        }
+        if (movementType === 'missingMovement') {
+            return formatMessage(MESSAGES.missingMovement, { movement });
+        }
+        return '';
+    };
```
Comment (quang-le) [inline plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/Modals/CreateEditIncident.tsx]: Can be moved out of the component

File: plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/Modals/CreateEditIncident.tsx [hunk 5]
```diff
@@ -59,6 +200,31 @@ export const CreateEditIncident: FunctionComponent<Props> = ({
...
+
+    const getMovementLabel = (movementType: IncidentReportFieldType) => {
+        switch (movementType) {
+            case 'plainMovement':
+            case 'missingMovement':
+                return MESSAGES.vialsOut;
+            default:
+                return MESSAGES.movement;
+        }
+    };
```
Comment (quang-le) [inline plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/Modals/CreateEditIncident.tsx]: same same

File: plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/Modals/CreateEditIncident.tsx [hunk 6]
```diff
@@ -100,22 +277,77 @@ export const CreateEditIncident: FunctionComponent<Props> = ({
...
+                        />
+                        <Typography variant="body2">
+                            {getMovementDescription(
+                                currentMovementType,
+                                formik.values.movement,
+                            )}
+                        </Typography>
+                    </Box>
+                )}
+                {currentMovementType && currentMovementType === 'inventory' && (
```
Comment (quang-le) [inline plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/Modals/CreateEditIncident.tsx]: `currentMovementType` cannot be equal to `'inventory'` if it's falsy 😉

File: plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/Table/columns.tsx [hunk 7]
```diff
@@ -220,6 +220,14 @@ export const useIncidentTableColumns = (
...
                         : textPlaceholder,
             },
+            {
+                Header: formatMessage(MESSAGES.title),
+                accessor: 'title',
+                id: 'title',
+                sortable: true,
+                Cell: settings =>
+                    settings.row.original.title || textPlaceholder,
+            },
```
Comment (quang-le) [inline plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/Table/columns.tsx]: You could use `BreakWordCell` here

**Review by quang-le — CHANGES_REQUESTED** [review]
Some movements are not translated in the usable/unusable tables:
Screenshot 2024-08-23 at 11 56 03

Otherwise nothing really blocking front-end side

Changes should be made in the backend for the movements that remove usable vials without adding the equivalent unusable vials.
I don't know if it's ok to merge the front-end changes without the backend fix, but without it, such movements are not shown in the main tables :

https://github.com/user-attachments/assets/6acbb5d6-1541-4c0d-8855-48922f04cd38

Jira ticket created for this issue: https://bluesquare.atlassian.net/browse/POLIO-1655

Adjacent to this PR, the table of incident reports is not very legible, since the columns "usable vials" and unusable vials show In and OUT movements in the same column:

Screenshot 2024-08-23 at 10 21 07

File: plugins/polio/js/src/domains/VaccineModule/StockManagement/hooks/useGetMovementDescription.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+    return (movementType: IncidentReportFieldType, movement: number) => {
+        if (movementType === 'plainMovement') {
+            return formatMessage(MESSAGES.plainMovement, { movement });
+        }
+        if (movementType === 'missingMovement') {
+            return formatMessage(MESSAGES.missingMovement, { movement });
+        }
+        return '';
+    };
+};
```
Comment (quang-le) [inline plugins/polio/js/src/domains/VaccineModule/StockManagement/hooks/useGetMovementDescription.ts:17]: That should be `useCallback` no?

**Review by quang-le — APPROVED** [review]

File: plugins/polio/js/src/domains/VaccineModule/StockManagement/hooks/useGetMovementDescription.ts (hunk 8 above)
Comment (beygorghor) [reply plugins/polio/js/src/domains/VaccineModule/StockManagement/hooks/useGetMovementDescription.ts:17]: This is o, fact returning a string, so not necessary i would say

**Pull request merged**